app/Jobs/FetchAd.php

```diff
@@ -9,108 +9,114 @@
+        $this->since = $since;
+        $this->source = Repo::getRecords('sources', ['*'], ['name' => $this->sourceName])->first();
+        $this->lastFetch = Repo::getRecords('fetches', ['*'], ['source_id' => $this->source->id, 'completed_at' => null])
+            ->orderByDesc('created_at')->first();
+        $this->adFetcher = $adFetcher;
+        $this->adItemInserter = $adItemInserter;
+        $this->adPicTransferrer = $adPicTransferrer;
     }
 
     /**
      * Execute the console command.
-     * @param AdFetcher $adFetcher
-     * @param AdItemInserter $adItemInserter
-     * @param AdPicTransferrer $adPicTransferrer
      */
-    public function handle(AdFetcher $adFetcher , AdItemInserter $adItemInserter , AdPicTransferrer $adPicTransferrer)
+    public function handle()
     {
-        $this->adFetcher = $adFetcher;
-        $this->adItemInserter=$adItemInserter;
-        $this->adPicTransferrer=$adPicTransferrer;
-
         /** @var stdClass $source */
-        $source = Repo::getRecords('sources' , ['*'], ['name'=>$this->sourceName])->first();
-        if(isset($source)){
-            [$donePages, $failedPages] = $this->fetch($source);
+        if (isset($this->source)) {
+            $donePages = $this->fetch();
+        } else {
+            Log::error('source is not set with name: ' . $this->sourceName);
+            throw new Exception('source is not set with name: ' . $this->sourceName);
         }
     }
 
     /**
-     * @param stdClass $source
-     * @return array
+     * @return int
      */
-    private function fetch(stdClass $source): array
+    private function fetch(): int
     {
-        $fetchUrl = (new SourceFetchUrlGenerator($source))->generateUrl();
+        $fetchUrl = (new SourceFetchUrlGenerator($this->source, $this->since))->generateUrl();
 
-        if(is_null($fetchUrl)) {
-            return [0, 0];
+        if (is_null($fetchUrl)) {
+            return 0;
         }
 
-        $failedPages = 0;
         $donePages = 0;
 
         do {
             $counter = 0;
-            [$fetchDone , $items , $currentPage , $nextPageUrl , $lastPage, $resultText] = $this->adFetcher->fetchAd($fetchUrl);
-            if(!$fetchDone){
-                $failedPages++;
-                continue;
+            [$fetchDone, $items, $currentPage, $nextPageUrl, $lastPage, $resultText] = $this->adFetcher->fetchAd($fetchUrl);
+            if (!$fetchDone) {
+                Log::error('response status code is not 200 | request url: ' . $fetchUrl);
+                throw new Exception('response status code is not 200 | request url: ' . $fetchUrl);
             }
 
             if (empty($items)) {
-                continue;
+                Log::error('response data is null | request url: ' . $fetchUrl);
+                throw new Exception('response data is null | request url: ' . $fetchUrl);
+            }
+            $this->storeItems($this->source, $items, $counter);
+
+            if ($currentPage === 1) {
```

Comment:
refactor: 
1- remove do while loop 
2- create fetch with model if first request was success
3- update fetch in for loop if request was success